### PR TITLE
feat: support custom OpenAI-compatible providers, per-model context sizes, dynamic context meter, and UI fixes

### PR DESCRIPTION
**Pull request by deathbypain**
#### Summary
This PR adds support for custom and local OpenAI-compatible LLM endpoints (such as `llama-swap`, `llama-server`, `Ollama`, `vLLM`, and custom relays), enables per-model context limit configuration, dynamically updates the chat context pressure meter when switching models, and resolves layout, hover, and localization issues in the model/mode picker UI.

---

#### 1. Custom & Local OpenAI-Compatible Provider Support
- **Keyless Provider Configuration:** Relaxed API key validation to allow optional/empty keys for local inference engines (e.g. `http://localhost:8080/v1`) that do not require authentication.
- **Provider & Model Catalog Discovery:** Custom providers configured via Connection Settings are fully surfaced in the webview model picker alongside official DeepSeek and relay routes.

#### 2. Per-Model Context Window Configuration
- **`model:size` Syntax Support:** Added the ability to specify custom context window limits per model in the provider's model list (e.g., `gemma-4-12b:32k`, `llama-3.3-70b:128k`, `qwen:64k`).
- **Parsing & Persistence:** Supports standard unit suffixes (`k`, `m`, `b` or raw numbers) with input validation, preserves custom overrides across provider profile edits, and cleans up stale model entries.

#### 3. Dynamic Context Pressure Recalculation
- **Live Meter Updates:** Exposed `contextWindow` on `ModelView` and `ModelConfigurationOption`.
- **Instant UI Refresh:** Selecting a different model in the model picker instantly recalculates and repaints the context pressure meter (`Xk / Yk - Z% used`) in the composer toolbar without requiring a reload or session restart.

#### 4. Model Picker UI & Layout Polish
- **Fixed Height & Single Scroll Viewport:** Stabilized the model picker popup dimensions so switching between providers with varying model counts doesn't resize the dialog or push it off-screen.
- **Hover Highlight Alignment:** Fixed CSS line-height and padding inconsistencies that caused the hover selection box to straddle multiple model rows.

#### 5. Localization for Built-In DSH Agent Modes
- **English Default for Shipped Presets:** Resolved an issue where built-in agent presets (`standard`, `ptc`, `minimal`, `cordis`) displayed raw Chinese titles and descriptions even in English locales.
- **`ptc` / `code` Alias Resolution:** Mapped the backend's `ptc` preset identifier to the extension's `code` catalog option and icon. Built-in presets now translate via `vscode.l10n.t`, while user-authored custom presets pass through unlocalized.

---

### **Commit History**

1. `536cf8b` — **feat: support custom openai-compatible providers and per-model context sizes**
   - Config validation, connection settings service & mapping, webview UI inputs, README documentation, and unit tests.
2. `0e050ff` — **fix(ui): stabilize model picker layout and model list scrolling**
   - CSS adjustments for fixed popup height, single scrolling container, and centered item hover states.
3. `a7c33c2` — **feat: dynamically update context pressure meter on model selection**
   - Workbench state, gateway capacity lookups, composer adapter/component live context calculation, and test coverage.
4. `c40e28d` — **fix(i18n): localize built-in DSH mode names in model picker**
   - Domain helper for localized preset display, gateway preset refresh mapping, icon matching, and test suite.

---

### **Testing & Verification**

- **Type Checking:** `check-types` passes cleanly (`tsc --noEmit`).
- **Test Suite:** Full test suite passes (771 tests across 93 test files), including new unit tests for:
  - `test/connection-settings-models-field.test.ts`
  - `test/connection-settings-service.test.ts`
  - `test/context-pressure.test.ts`
  - `test/agent-preset-display.test.ts`
  - `test/composer-configuration-adapter.test.ts`
- **Manual Verification:** Tested in Extension Development Host against local `llama-swap` endpoint, verifying model discovery, per-model `:contextSize` parsing, dynamic context meter recalculation on model switch, fixed-height popup scrolling, and English DSH mode labels.

---

AI disclosure: Github Copilot was used to generate the majority of the code. I, the human, directed the assistant for each individual feature and fix, manually approved every tool call, and reviewed the code (though I will admit my experience with typescript specifically is limited). I tested the results comprehensively on my own system, running llama-server via llama-swap with a selection of local models.

## Summary by CodeRabbit

- **New Features**
  - OpenAI-compatible providers support local endpoints without API keys.
  - Connection testing can import available model IDs, which can also be entered manually.
  - Model IDs support context-window overrides such as `:32k`; unspecified models use known capacities or a 256K default.
  - Context-window settings improve context indicators and model-switching accuracy.
  - Built-in preset names and descriptions are localized consistently.

- **Bug Fixes**
  - Improved model-panel scrolling and prevented nested scrollbars.
  - Duplicate model IDs preserve the latest context-window override.

### CODE REVIEW

**Comment by deathbypain** [comment]
I'm working on resolving some of code rabbit's suggestions, will update hopefully within a few days

File: src/services/connection-settings/mapping.ts [hunk 1]
```diff
@@ -89,6 +92,26 @@ export function normalizeRelayModels(models: readonly string[] | undefined): rea
...
+  contextWindows: Readonly<Record<string, number>> | undefined,
+  models: readonly string[],
+): Readonly<Record<string, number>> {
+  if (contextWindows === undefined) return {}
+  const known = new Set(models)
+  const normalized: Record<string, number> = {}
+  for (const [id, value] of Object.entries(contextWindows)) {
+    if (!known.has(id)) continue
+    if (!Number.isFinite(value) || value <= 0) continue
+    normalized[id] = Math.round(value)
```
[the remark this answers is not in the document]
Comment (deathbypain) [reply src/services/connection-settings/mapping.ts]: Addressed in commit 19b2371

File: src/webview/composer-configuration/component.ts [hunk 2]
```diff
@@ -102,6 +104,10 @@ class ComposerConfigurationDom implements ComposerConfigurationComponent {
     return this.store.snapshot()?.selection
   }
 
+  stagedModel(): ModelConfigurationOption | undefined {
+    return this.store.snapshot()?.model
+  }
```
[the remark this answers is not in the document]
Comment (deathbypain) [reply src/webview/composer-configuration/component.ts:114]: Addressed in commit 5b03c8b

**Review by skymecode — COMMENTED** [review]
> ### 🤖 This review is 100% AI-generated
>
> **No human reviewed this code before this comment was posted.** The code reading, the execution
> checks and the wording below were all produced by an AI agent.
>
> | | |
> |---|---|
> | **Model** | DeepSeek V4 Flash (`deepseek-v4-flash`) |
> | **Agent** | DSH (DeepSeek Harness) coding agent |
> | **Input** | Full PR diff at head `8fdb66d` (23 files, +525/−52), plus a local checkout of the PR branch for execution-level checks |
> | **Human oversight** | None. Treat every claim below as unverified until a human confirms it — I can be wrong, and the blocker is the one I would re-check first. |

Thanks for the PR. Support for local/OpenAI-compatible endpoints is a real gap, and the
`:contextSize` syntax plus the live context meter are the right shape for it. Findings are ordered
by severity.

---

## 🔴 Blocker — `parseModelsField()` silently truncates model ids that contain `:`

`src/webview/connection-settings/component.ts`

```ts
const separator = token.lastIndexOf(':')
const id = (separator < 0 ? token : token.slice(0, separator)).trim()
if (id === '') continue
ids.push(id)                                  // ← pushed before the suffix is validated
if (separator < 0) continue
const size = parseContextWindow(token.slice(separator + 1).trim())
```

I extracted this function verbatim from the PR head and executed it:

| Input | Result |
|---|---|
| `llama3.2:3b` | `{ ids: ["llama3.2"], contextWindows: {} }` ❌ tag dropped |
| `qwen2.5-coder:7b` | `{ ids: ["qwen2.5-coder"], contextWindows: {} }` ❌ |
| `deepseek-r1:latest` | `{ ids: ["deepseek-r1"], contextWindows: {} }` ❌ |
| `meta-llama/llama-3.3-70b-instruct:free` | `{ ids: ["meta-llama/llama-3.3-70b-instruct"] }` ❌ |
| `gemma-4-12b:32k` | `{ ids: ["gemma-4-12b"], contextWindows: { "gemma-4-12b": 32768 } }` ✅ |

**Why this matters for this PR in particular:**

1. **It hits the PR's own headline use cases.** Ollama / llama.cpp model ids are `name:tag` — the PR
   description explicitly lists `llama-server`, `llama-swap` and `Ollama`. `:3b`, `:7b`, `:latest`
   are part of the model name, not a size suffix. Note that `model-capacity.ts` in this same repo
   already treats `:free` / `:latest` as part of the id, so the two paths disagree.
2. **The auto-import path amplifies it.** `renderTestResult` writes the endpoint's advertised ids
   into the field verbatim (`llama3.2:3b`). Clicking **Apply** then re-parses the field and stores
   `llama3.2` in the provider profile — a model id the endpoint does not serve, so the failure only
   surfaces later as a request error.
3. **This is a regression, not just a missing feature.** On `main` the field is
   `models.value.split(/[,，\s]+/)`, which preserves `llama3.2:3b`. An existing Ollama user who opens
   Connection settings and clicks Apply now gets a silently rewritten model id.

**Suggested fix** — only treat the suffix as a size when it actually parses as one; otherwise keep
the whole token as the id:

```ts
const separator = token.lastIndexOf(':')
const size = separator < 0 ? undefined : parseContextWindow(token.slice(separator + 1).trim())
if (size === undefined) { ids.push(token); continue }   // the colon belongs to the model name
const id = token.slice(0, separator).trim()
if (id === '') continue
ids.push(id)
contextWindows[id] = size
```

`lastIndexOf` still makes `llama3.2:3b:32k` work correctly with this change.

Related: `test/connection-settings-models-field.test.ts` currently locks in the lossy behaviour
(`model-a:not-a-number` → `ids: ['model-a']`), so that expectation needs to move together with the
fix. Silently swapping in a *different, existing* model (`llama3.2` resolves to `:latest`) is worse
than surfacing the full token, which fails loudly.

---

## 🟡 Should fix

**1. Capacity-table backfill is persisted as if the user had set it** — `src/services/connection-settings/mapping.ts`

`modelContextWindowsField()` reads back the `contextWindow` of *every* model entry, including the
values `relayModels()` backfills from the bundled capacity table. Those then render in the Model IDs
field as `deepseek-v4-flash:1m`, and the next Apply writes them into `modelContextWindows` as
explicit user overrides — permanently pinning a value the user never typed, so a later correction to
the capacity table can no longer reach that provider. The PR's own test asserts this:
the input specifies no sizes, yet the expected state contains
`modelContextWindows: { 'deepseek-v4-flash': 1_000_000, 'deepseek-v4-pro': 1_000_000 }`.
Consider keeping "what the user typed" separate from "what is in effect".

**2. Context-window precedence overrides the harness-reported value** — `src/gateway/harness-gateway-service.ts`

```ts
const effectiveContextWindow = this.resolvedContextWindow(currentProvider, currentModel)
  ?? rawContextPressure?.contextWindow
```

`resolvedContextWindow()` is *user override → bundled capacity table → undefined*, so the hard-coded
table also wins for the official DeepSeek provider, where the harness already reports the
authoritative window. The table is a snapshot, so a server-side change would leave the meter stale.
For custom relays the table is the right fallback (the adapter reports the 256K default) — it is
only the precedence for non-custom providers that looks inverted.

**3. `deepseekHarness.model` still declares a two-value enum** — `package.json`

This PR widens `HarnessConfiguration.model` from `ModelId` to `string`, which is correct for custom
providers — but the user-facing setting is still
`enum: ["deepseek-v4-flash", "deepseek-v4-pro"]`. `setModelId()` writes custom ids into it, so the
Settings UI will flag them as invalid values. (Persisting ids outside the whitelist predates this
PR, but this PR is what makes custom models a first-class feature, so it seems worth closing here.)

**4. The Chinese hint text is malformed** — `l10n/bundle.l10n.zh-cn.json`

The new `providerModelsHint` translation has 2 × `（` but only 1 × `）`, and the clause reads
ungrammatically:

> …已知模型会使用内置容量表中的值（如果已知，未知模型否则默认按 256K 处理。点击“测试连接”可自动填充 ID。

Suggested: 「…大多数本地端点不会上报该信息：已知模型沿用内置容量表中的值，未知模型默认按 256K 处理。点击“测试连接”可自动填充 ID。」 `test/localization.test.ts` only checks key
presence, so this passes CI but is visible to users.

**5. `npm run lint` fails** — `src/config/configuration.ts`

After switching to `configuredModel()`, the `modelId` import is no longer used. Verified by running
ESLint against the file at PR head:

```
7:3  error  'modelId' is defined but never used  @typescript-eslint/no-unused-vars
```

`tsc --noEmit` does not catch this (`noUnusedLocals` is not enabled), so it would only show up when
lint runs. Trivial to drop.

---

## 🔵 Minor / hardening

- **Keyless remote providers**: after removing the empty-key check, a *remote* base URL can also be
  saved without a key; the failure is deferred to request time. A non-blocking hint for non-loopback
  hosts would help — local endpoints legitimately need no key.
- **Unquoted YAML interpolation** — `src/runtime/runtime-overlay.ts` writes `model: ${configuration.model}`,
  while the neighbouring `provider` uses `JSON.stringify(...)`. That was safe while `model` was a
  closed two-value union; now that it can be any string sourced from a user or an endpoint's
  `/v1/models`, an id containing YAML-special characters (`#`, a leading `*`, a newline) could
  produce a malformed overlay. Cheap to make consistent.
- **`ptc` / `code` identity**: the alias is display-only. The roster reports `ptc`, while
  `AGENT_PRESET_OPTIONS` and the setting enum use `code`, so `setAgentPresetIfKnown('ptc')` will not
  persist. Out of scope for this PR, but worth confirming it is intended.
- **Fixed `height: 280px`** on `.configuration-provider-layout`: the panel keeps
  `max-height: min(72vh, 480px)` and a scrollable middle row, so I did not find a clipping path — but
  per this repo's webview guidelines it is worth re-checking on Windows and in the narrowest sidebar.

---

## ✅ What looks good

- **Defence in depth on sizes**: zero / negative / NaN are rejected at three layers — the webview
  input validator, `normalizeModelContextWindows()` and `parseContextWindow()`.
- `normalizeModelContextWindows()` drops stale overrides for ids the provider no longer exposes.
- `localizedPresetDisplay()` only rewrites `trust: 'system'` presets and passes user-authored ones
  through untouched — the right boundary.
- The key relaxation only affects provider *creation*; editing with a blank key still means "keep the
  existing credential", so that semantic is preserved.
- The follow-up commits caught two real edge cases on their own (`stagedModel()` guessing while Auto
  is active; context sizes rounding to zero).
- Test coverage is solid: 40+ root-level filenames × 4 location suffixes, cross-platform path
  spellings, external-reference exclusions, `k`/`m` and full-width commas.
- Disclosing the extent of AI involvement and the manual testing performed in the PR description is
  much appreciated.

---

## How this was checked

Read the full diff; fetched the PR head locally; executed `parseModelsField()` verbatim from that
head in Node; ran ESLint against the PR version of `configuration.ts`; compared the field-parsing
behaviour against `main`. **The full test suite was not run against the PR head** — the author's
reported results (771 tests / 93 files) were taken as given.

Happy to expand on any of these; the blocker is the only item I would consider release-blocking.

File: src/webview/connection-settings/component.ts [hunk 3]
```diff
@@ -227,3 +238,51 @@ function required<T extends HTMLElement>(document: Document, id: string): T {
...
+/** Parses a raw token count or a `k`/`m`-suffixed size into tokens; undefined when invalid. */
+function parseContextWindow(raw: string): number | undefined {
+  const match = /^(\d+(?:\.\d+)?)\s*([km])?$/iu.exec(raw)
+  if (match === null) return undefined
+  const value = Number(match[1])
+  if (!Number.isFinite(value)) return undefined
+  const unit = match[2]?.toLowerCase()
+  const multiplier = unit === 'k' ? 1024 : unit === 'm' ? 1024 * 1024 : 1
+  const tokens = Math.round(value * multiplier)
+  return tokens > 0 ? tokens : undefined
```
[the remark this answers is not in the document]
Comment (deathbypain) [reply src/webview/connection-settings/component.ts]: addressed in 3516f58

**Comment by skymecode** [comment]
> ### 🤖 This review is AI-generated
>
> The code analysis, execution checks, and wording below were produced by an AI agent.
>
> | | |
> |---|---|
> | **Model** | GPT-6 |
> | **Agent** | Codex |
> | **Input** | PR head `defcbfeac11323598dc1285802533e67f1b4b511`, an isolated local checkout, and the lockfile-installed `@deepseek-ai/dsh-llm-pi-ai@0.1.5-alpha.1` runtime |
> | **Human oversight** | The repository owner requested this follow-up. The technical findings and reproduction were performed by the AI agent; no independent human verification is claimed. |

Thanks for the follow-up fixes. I rechecked the latest head and found a separate **P1 blocker in the keyless-provider path**. This comment is limited to that issue.

---

## 🔴 P1 — Keyless providers can be saved, but requests still fail with `MISSING_CREDENTIAL`

Location: [`src/services/connection-settings-service.ts`, lines 163–178](https://github.com/skymecode/deepseek-harness-for-vscode/blob/defcbfeac11323598dc1285802533e67f1b4b511/src/services/connection-settings-service.ts#L163-L178), together with `deepSeekRelayProfile()` in `src/services/connection-settings/mapping.ts`.

Removing the empty-key validation allows the form to save a new local provider. However, the saved profile still unconditionally contains an `apiKeyEnv` reference, while `credentialsSet()` is skipped when the key is empty.

For example, creating `Local Llama` with an empty API key produces a profile containing:

```json
{
  "displayName": "Local Llama",
  "baseURL": "http://127.0.0.1:8080/v1",
  "api": "openai-completions",
  "apiKeyEnv": "PROVIDER_LOCAL_LLAMA_API_KEY"
}
```

No credential is stored for `PROVIDER_LOCAL_LLAMA_API_KEY`. In the pinned DSH runtime, an explicit but unresolved `apiKeyEnv` is a hard error: `resolveApiKey()` throws before the HTTP request is sent.

### Execution-level reproduction

I passed the profile generated by this PR's actual `deepSeekRelayProfile()` into the installed DSH plugin's exported `Config` and `apply()` functions. The test context provided an empty credential store; the registered adapter and discovery handler were the real runtime implementations.

| Operation | Result |
|---|---|
| Start an adapter stream for the saved local provider | `MISSING_CREDENTIAL` |
| Probe the saved provider with its provider id and explicit `baseURL` | `MISSING_CREDENTIAL` |

Both returned:

```text
llm-pi-ai: no credential for provider route "local-llama"; its profile resolves PROVIDER_LOCAL_LLAMA_API_KEY, which is not set
```

No live inference server was needed for this reproduction: execution fails during credential resolution, before contacting the endpoint. This is an adapter-level reproduction, not a claim of a successful end-to-end local-model run.

### Why this blocks the advertised feature

A fresh user can leave the key empty and save the provider, but cannot send a prompt through it. Connection discovery before saving does not establish that the saved route will work, because the saved route subsequently resolves the credential reference above.

The new `creates a custom provider without requiring an API key` test only checks profile persistence and that no credential was written. It does not exercise the saved profile through the runtime, so it passes while the request path remains broken.

### Suggested fix and regression coverage

- Represent an intentionally keyless provider without leaving a required credential reference unresolved, and verify the complete pi-ai request path. Removing `apiKeyEnv` alone should not be assumed sufficient without testing the transport's own authentication requirements.
- Preserve the existing edit behavior: leaving the key field blank on an already authenticated provider must continue to keep its stored credential.
- Add a regression test that creates a provider with an empty key, then uses the saved configuration for both model discovery and a completion against a local mock OpenAI-compatible endpoint. Assert that requests reach the endpoint instead of failing during credential resolution.

### How this was checked

- Installed the PR's locked dependencies with `npm ci` in an isolated checkout.
- `npx tsc --noEmit`: passed.
- `npx vitest run`: **773 passed, 15 skipped**.
- `node esbuild.mjs --production`: passed.
- Ran the targeted runtime reproduction described above; both paths failed with `MISSING_CREDENTIAL`.

Please fix this saved-provider runtime path before merging the keyless-provider feature.
